### PR TITLE
📊 Add engine events to contact history

### PR DESCRIPTION
**Pull request by rowanseymour**
Figured I'd start simple and just add one event type from sessions...

Captura de Pantalla 2019-09-26 a la(s) 17 03 43

### CODE REVIEW

File: locale/en_US/LC_MESSAGES/django.po [hunk 1]
```diff
@@ -5562,6 +5562,14 @@ msgstr ""
 msgid "Call Started"
 msgstr ""
 
+#, python-format
+msgid ""
+"\n"
+"          Run result <i>%(result_name)s</i> updated to <i>%(category)s</i>\n"
```
Comment (rowanseymour) [inline locale/en_US/LC_MESSAGES/django.po]: this seem right for a piece of text with some html? also is all that whitespace significant?

File: temba/contacts/models.py [hunk 2]
```diff
@@ -799,59 +799,96 @@ def get_scheduled_messages(self):
...
-            [{"type": "call", "time": c.created_on, "obj": c} for c in calls],
+        events = chain(
+            [{"type": "msg_created", "created_on": m.created_on, "obj": m} for m in msgs_out],
+            [{"type": "msg_received", "created_on": m.created_on, "obj": m} for m in msgs_in],
+            [{"type": "flow_entered", "created_on": r.created_on, "obj": r} for r in started_runs],
+            [{"type": "flow_exited", "created_on": r.exited_on, "obj": r} for r in exited_runs],
+            [{"type": "channel_event", "created_on": e.created_on, "obj": e} for e in channel_events],
+            [{"type": "campaign_fired", "created_on": f.fired, "obj": f} for f in campaign_events],
+            [{"type": "webhook_called", "created_on": r.created_on, "obj": r} for r in webhook_results],
+            [{"type": "call_started", "created_on": c.created_on, "obj": c} for c in calls],
```
Comment (rowanseymour) [inline temba/contacts/models.py]: making these look more like real session events so it's easier to mix the two

**Comment by rowanseymour** [comment]
Current look of things...

Captura de Pantalla 2019-09-30 a la(s) 15 26 11


* Group and label names are links to their filter pages
* New sessions get their own pseudo-events to make space to describe the trigger - the only we see these

**Comment by ericnewcomer** [comment]
This is going to be awesome. Why did we wait again??

One immediate thought here is that maybe we should use a similar treatment as the editor (using font weight for the things being changed instead of italics).

Is the "Field" example actually named "Field"?

**Comment by rowanseymour** [comment]
@ericnewcomer that works - using `#666` to match the field values above. Definitely easier to read.

Field name being blank was an error

Captura de Pantalla 2019-09-30 a la(s) 15 42 08

**Comment by nicpottier** [comment]
Looking good.

For a few things we have icons we use elsewhere that we may want to match. Specifically the editor uses an icon for URNs that we may want to use.

All the heavier weight feels a bit jarring to me for some reason, but I'm sure I will get used to it and it probably is fine. Woot!

**Comment by ericnewcomer** [comment]
Agreed, something lighter than bold would be nicer here, but we can tweak later.

**Comment by rowanseymour** [comment]
@nicpottier are you looking at the old editor? In the new editor there's only a few icons I can see for labels, groups and flows - I think the icons I've taken from the icomoon stuff are pretty close to those

**Comment by nicpottier** [comment]
It is, you will see them if you go to an update contact action and look at
the icon next to URNs. It's like a circle surrounding a dot.

**Comment by norkans7** [comment]
Any way to show the same events if the contacts if updated on the web?

**Comment by nicpottier** [comment]
We don't really have that data yet but if we do:
  https://github.com/rapidpro/rapidpro/issues/1016

Then agree that would fit in nicely. Maybe that's a task for you Norbert to add a model to track user edits of groups / fields / urns.

File: locale/en_US/LC_MESSAGES/django.po (hunk 1 above)
Comment (nicpottier) [reply locale/en_US/LC_MESSAGES/django.po]: I think the whitespace is just HAML garbage and not significant but difficult to remove entirely. To be perfectly honest I don't know what the best practice is when it comes to markup here. You can break them up, but then that is harder for the translator to know context, or you can have it all munged like this and expect them to include the tags. I personally prefer the latter but like I said no idea on best practice.

**Comment by nicpottier** [comment]
Looks great. I think we will likely have to iterate on this in the future either to help explain why events disappear at some point (or better yet to make them not do that) but this is a big win as is.

**Comment by rowanseymour** [comment]
There's one thing that doesn't quite work still.. and that's that I added a pseudo-event called session_started to display stuff from the trigger about why the session was triggered... but that ends up being a bit repetitious cuz we already list campaign event fires and channel events... so think I'm going to rip that bit out and we can revisit whether can provide more information about why flows were started.

**Pull request merged**